### DISCUSSION TITLE
How to switch a schema property between optional and required at run time

### DISCUSSION THREAD

**Discussion by smcardle** [colinhacks/zod#1789] [Q&A] [answered]
Hi All.
I'm using Zod with react-hook-form and have a need to be able to dynamically switch properties in a zod schema between optional and required based on the value of another property in the schema.
Once a property is set as optional, I cannot set the field as required in my form as RHF will then not consider the zod schema and reverts to native form handling for required fields, thus loosing the benefit of the schema validation for creating FHF errors.

I also use the same schema in XState for the context and cannot define all of the properties as required otherwise I cannot meet the rules i.e. this is REQUIRED but only if property X=valueY

Steve

**Comment by JacobWeisenburger** [comment]
Please send some example code.

**Comment by smcardle** [comment]
Thanks for getting back to me.

I'm not sure what you mean by some "example code" but here is a specific schema

```ts
export const entitled = ['entitled','notentitled'] as const; 
export const citizenship = ['citizen', 'noncitizen'] as const;
export const eligibility = ['residentviza', 'australian', 'workviza', 'interimviza', 'refugee', 'under18', 'aidprogramme', 'teaching', 'scholarship'] as const;
export const FormData = z.object({
  entitled: z.enum(entitled, {errorMap: (issue, ctx) => {
    return {message: 'Please select one of the above options'};
  },}),
  citizenship: z.enum(citizenship).optional(),
  eligibility: z.enum(eligibility).optional(),
});
export type TFormData = z.infer<typeof FormData>;
```

As can be seen the citizenship and eligibility properties are defined here optional.

The rules for this are 
1. If entitled is equal to notentitled then both citizenship and eligibility are not required i.e. optional
2. If entitled is equal to entitled then citizenship is required 
3. If citizenship is equal to citizen then eligibility is not required
4. If citizenship is equal to noncitizen then eligibility is required.

I have several instances of similar rules within the code base like this along with others such as:
One step in the form has Mobile Number required with an optional Home Number. However, there is a check box for the user to select they do not have a Mobile Number and once selected the rule changes to Mobile Number is optional and Home Number is required.

Basically Rules dictated by value of other properties previously set.

For react-hook-form to properly process these they have to be optional or required (as above) at the time RHF validates. Setting the component in the form to required makes RHF just use the native browser validation for those fields and ignore any validation rules in Zod via the Zod Resolver

I also use the same schema as the context for XState. 
This means it's NOT possible to set all the fields in the above schema to say required, as then XState's Typescript checking fails when trying to set any of the above optional fields to undefined.


So the question really is.... How can I satisfy all of these rules with the schema and in ALL of my cases, being able to switch a field between required and mandory in the schema at runtime would do this.

Steve

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const entitledEnum = {
    entitled: 'entitled',
    notentitled: 'notentitled',
} as const
const citizenshipEnum = {
    citizen: 'citizen',
    noncitizen: 'noncitizen',
} as const
const eligibility = [
    'residentviza', 'australian', 'workviza',
    'interimviza', 'refugee', 'under18',
    'aidprogramme', 'teaching', 'scholarship'
] as const

const schema = z.union( [
    z.object( {
        entitled: z.literal( entitledEnum.entitled ),
        citizenship: z.literal( citizenshipEnum.citizen ),
        eligibility: z.enum( eligibility ).optional(),
    } ),
    z.object( {
        entitled: z.literal( entitledEnum.entitled ),
        citizenship: z.literal( citizenshipEnum.noncitizen ),
        eligibility: z.enum( eligibility ),
    } ),
    z.object( {
        entitled: z.literal( entitledEnum.notentitled ),
        citizenship: z.literal( citizenshipEnum.citizen ).optional(),
        eligibility: z.enum( eligibility ).optional(),
    } ),
    z.object( {
        entitled: z.literal( entitledEnum.notentitled ),
        citizenship: z.literal( citizenshipEnum.noncitizen ).optional(),
        eligibility: z.enum( eligibility ),
    } ),
] )
```

**Reply by smcardle** [reply]
I've actually just fixed this one issue with the following (leaving off the actual enums)
I'm not happy with 3 x the same schema as that is a lot of duplication considering this is one of the simplets multi required / optional rules

```ts
export const FormData = z
  .object({
    entitled: z.enum(entitled).optional(),
    citizenship: z.enum(citizenship).optional(),
    eligibility: z.enum(eligibility).optional(),
  })
  .superRefine((val, ctx) => {
    // notentitled is a valid option on it's own no need to check other conditions
    if(val.entitled === 'notentitled') return true;

    if(!val.entitled) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['entitled'],
        message: 'You must select one of the options displayed',
      });
      return true;
    }
    else if(!val.citizenship) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['citizenship'],
        message: 'You must select one of the options displayed',
      });
      return true;
     }
     else if(val.citizenship === 'noncitizen' && !val.eligibility) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['eligibility'],
        message: 'You must select one of the options displayed',
      });
      return true;
     }
  });
export type TFormData = z.infer<typeof FormData>;

```
Your answer would spiral out of control just by adding another rule to the schema.

Still not entirely happy with this resolution either as I really wanted to keep any logic out of the context as it should really be implemented as a guard in my XState implementation.
However, this kepts it in just one place and allows me to remove the guard in XState resulting in the logic being in just one place, even if it's in the wrong place.

New rules would just require more conditional checks in the superRefine rather than X number of more unions for effectively the same schema

I feel somewhat better about this solution as I did not want to go back to Yup and use the when(..) clause which I feel is even worse than this solution.

It would be so much simpler if there was a way to just add mandatory / optional to a property at run time.
To me this is an issue to some extent in all of the run time schema validation libraries.

This is also an issue with Form libraries such as react-hook-form. The natural way for devs to make a field mandatory is to add required to the html element. But unfortunately trying this make react-hook-form resort to native validation and ignores the Zod schema validation.

If there was a "library" way to add extra validation rules at runtime, then react-hook-form could modify the schema at runtime to also consider required/min/max etc entered by the dev, basically overriding just the part i.e. required defined by them.

Doing things the current way means the UI / UX dev ALSO needs to be very aware of the validation schema as well

**Reply by JacobWeisenburger** [reply]
I'm glad you were able to get something working.

**Comment by JacobWeisenburger** [comment, marked as answer]
> I've actually just fixed this one issue with the following (leaving off the actual enums) I'm not happy with 3 x the same schema as that is a lot of duplication considering this is one of the simplets multi required / optional rules
> 
> ```ts
> export const FormData = z
>   .object({
>     entitled: z.enum(entitled).optional(),
>     citizenship: z.enum(citizenship).optional(),
>     eligibility: z.enum(eligibility).optional(),
>   })
>   .superRefine((val, ctx) => {
>     // notentitled is a valid option on it's own no need to check other conditions
>     if(val.entitled === 'notentitled') return true;
> 
>     if(!val.entitled) {
>       ctx.addIssue({
>         code: z.ZodIssueCode.custom,
>         path: ['entitled'],
>         message: 'You must select one of the options displayed',
>       });
>       return true;
>     }
>     else if(!val.citizenship) {
>       ctx.addIssue({
>         code: z.ZodIssueCode.custom,
>         path: ['citizenship'],
>         message: 'You must select one of the options displayed',
>       });
>       return true;
>      }
>      else if(val.citizenship === 'noncitizen' && !val.eligibility) {
>       ctx.addIssue({
>         code: z.ZodIssueCode.custom,
>         path: ['eligibility'],
>         message: 'You must select one of the options displayed',
>       });
>       return true;
>      }
>   });
> export type TFormData = z.infer<typeof FormData>;
> ```
> 
> Your answer would spiral out of control just by adding another rule to the schema.
> 
> Still not entirely happy with this resolution either as I really wanted to keep any logic out of the context as it should really be implemented as a guard in my XState implementation. However, this kepts it in just one place and allows me to remove the guard in XState resulting in the logic being in just one place, even if it's in the wrong place.
> 
> New rules would just require more conditional checks in the superRefine rather than X number of more unions for effectively the same schema
> 
> I feel somewhat better about this solution as I did not want to go back to Yup and use the when(..) clause which I feel is even worse than this solution.
> 
> It would be so much simpler if there was a way to just add mandatory / optional to a property at run time. To me this is an issue to some extent in all of the run time schema validation libraries.
> 
> This is also an issue with Form libraries such as react-hook-form. The natural way for devs to make a field mandatory is to add required to the html element. But unfortunately trying this make react-hook-form resort to native validation and ignores the Zod schema validation.
> 
> If there was a "library" way to add extra validation rules at runtime, then react-hook-form could modify the schema at runtime to also consider required/min/max etc entered by the dev, basically overriding just the part i.e. required defined by them.
> 
> Doing things the current way means the UI / UX dev ALSO needs to be very aware of the validation schema as well
